File: frontend/lib/services/ocr_service.dart
import 'dart:io';
import 'package:dio/dio.dart';
import 'package:flutter/foundation.dart';

/// OCR识别类型枚举
enum OCRType {
  general,
  licensePlate,
  vin,
  invoice,
}

/// OCR识别结果模型
class OCRResult {
  final bool success;
  final String message;
  final Map<String, dynamic>? data;
  final String? error;

  OCRResult({
    required this.success,
    required this.message,
    this.data,
    this.error,
  });

  factory OCRResult.fromJson(Map<String, dynamic> json) {
    return OCRResult(
      success: json['success'] ?? false,
      message: json['message'] ?? '',
      data: json['data'],
      error: json['error'],
    );
  }
}

class OCRService {
  static const String _baseUrl = 'http://localhost:3000/api';
  final Dio _dio = Dio();

  /// 通用OCR文字识别
  Future<OCRResult> recognizeText(File imageFile, OCRType type) async {
    try {
      final formData = FormData.fromMap({
        'image': await MultipartFile.fromFile(
          imageFile.path,
          filename: imageFile.path.split('/').last,
        ),
        'type': _getTypeString(type),
      });

      final response = await _dio.post(
        '$_baseUrl/ocr/recognize',
        data: formData,
        options: Options(
          headers: {
            'Content-Type': 'multipart/form-data',
          },
        ),
      );

      return OCRResult.fromJson(response.data);
    } catch (e) {
      debugPrint('OCR识别失败: $e');
      return OCRResult(
        success: false,
        message: '识别失败: ${e.toString()}',
        error: e.toString(),
      );
    }
  }

  /// 车牌号识别
  Future<LicensePlateResult> recognizeLicensePlate(File imageFile) async {
    try {
      final formData = FormData.fromMap({
        'image': await MultipartFile.fromFile(
          imageFile.path,
          filename: imageFile.path.split('/').last,
        ),
      });

      final response = await _dio.post(
        '$_baseUrl/ocr/license-plate',
        data: formData,
        options: Options(
          headers: {
            'Content-Type': 'multipart/form-data',
          },
        ),
      );

      if (response.data['success']) {
        return LicensePlateResult.fromJson(response.data['data']);
      } else {
        throw Exception(response.data['message']);
      }
    } catch (e) {
      debugPrint('车牌识别失败: $e');
      throw Exception('车牌识别失败: ${e.toString()}');
    }
  }

  /// VIN码识别
  Future<VINResult> recognizeVIN(File imageFile) async {
    try {
      final formData = FormData.fromMap({
        'image': await MultipartFile.fromFile(
          imageFile.path,
          filename: imageFile.path.split('/').last,
        ),
      });

      final response = await _dio.post(
        '$_baseUrl/ocr/vin',
        data: formData,
        options: Options(
          headers: {
            'Content-Type': 'multipart/form-data',
          },
        ),
      );

      if (response.data['success']) {
        return VINResult.fromJson(response.data['data']);
      } else {
        throw Exception(response.data['message']);
      }
    } catch (e) {
      debugPrint('VIN码识别失败: $e');
      throw Exception('VIN码识别失败: ${e.toString()}');
    }
  }

  /// 发票识别
  Future<InvoiceResult> recognizeInvoice(File imageFile) async {
    try {
      final formData = FormData.fromMap({
        'image': await MultipartFile.fromFile(
          imageFile.path,
          filename: imageFile.path.split('/').last,
        ),
      });

      final response = await _dio.post(
        '$_baseUrl/ocr/invoice',
        data: formData,
        options: Options(
          headers: {
            'Content-Type': 'multipart/form-data',
          },
        ),
      );

      if (response.data['success']) {
        return InvoiceResult.fromJson(response.data['data']);
      } else {
        throw Exception(response.data['message']);
      }
    } catch (e) {
      debugPrint('发票识别失败: $e');
      throw Exception('发票识别失败: ${e.toString()}');
    }
  }

  /// 获取OCR配置信息
  Future<OCRConfig> getOCRConfig() async {
    try {
      final response = await _dio.get('$_baseUrl/ocr/config');

      if (response.data['success']) {
        return OCRConfig.fromJson(response.data['data']);
      } else {
        throw Exception(response.data['message']);
      }
    } catch (e) {
      debugPrint('获取OCR配置失败: $e');
      throw Exception('获取OCR配置失败: ${e.toString()}');
    }
  }

  /// 将OCR类型转换为字符串
  String _getTypeString(OCRType type) {
    switch (type) {
      case OCRType.general:
        return 'general';
      case OCRType.licensePlate:
        return 'license_plate';
      case OCRType.vin:
        return 'vin';
      case OCRType.invoice:
        return 'invoice';
    }
  }
}

/// 车牌识别结果
class LicensePlateResult {
  final String plateNumber;
  final String color;
  final double confidence;
  final Map<String, dynamic>? location;

  LicensePlateResult({
    required this.plateNumber,
    required this.color,
    required this.confidence,
    this.location,
  });

  factory LicensePlateResult.fromJson(Map<String, dynamic> json) {
    return LicensePlateResult(
      plateNumber: json['plateNumber'] ?? '',
      color: json['color'] ?? '',
      confidence: (json['confidence'] ?? 0).toDouble(),
      location: json['location'],
    );
  }
}

/// VIN码识别结果
class VINResult {
  final List<String> vins;
  final String fullText;
  final double confidence;

  VINResult({
    required this.vins,
    required this.fullText,
    required this.confidence,
  });

  factory VINResult.fromJson(Map<String, dynamic> json) {
    return VINResult(
      vins: List<String>.from(json['vins'] ?? []),
      fullText: json['fullText'] ?? '',
      confidence: (json['confidence'] ?? 0).toDouble(),
    );
  }
}

/// 发票识别结果
class InvoiceResult {
  final InvoiceInfo invoice;
  final RepairInfo? repairInfo;
  final String fullText;

  InvoiceResult({
    required this.invoice,
    this.repairInfo,
    required this.fullText,
  });

  factory InvoiceResult.fromJson(Map<String, dynamic> json) {
    return InvoiceResult(
      invoice: InvoiceInfo.fromJson(json['invoice'] ?? {}),
      repairInfo: json['repairInfo'] != null
          ? RepairInfo.fromJson(json['repairInfo'])
          : null,
      fullText: json['fullText'] ?? '',
    );
  }
}

/// 发票信息
class InvoiceInfo {
  final String invoiceType;
  final String invoiceCode;
  final String invoiceNum;
  final String invoiceDate;
  final String totalAmount;
  final String amountInWords;
  final String sellerName;
  final String purchaserName;
  final List<CommodityDetail> commodityDetails;

  InvoiceInfo({
    required this.invoiceType,
    required this.invoiceCode,
    required this.invoiceNum,
    required this.invoiceDate,
    required this.totalAmount,
    required this.amountInWords,
    required this.sellerName,
    required this.purchaserName,
    required this.commodityDetails,
  });

  factory InvoiceInfo.fromJson(Map<String, dynamic> json) {
    return InvoiceInfo(
      invoiceType: json['invoiceType'] ?? '',
      invoiceCode: json['invoiceCode'] ?? '',
      invoiceNum: json['invoiceNum'] ?? '',
      invoiceDate: json['invoiceDate'] ?? '',
      totalAmount: json['totalAmount'] ?? '',
      amountInWords: json['amountInWords'] ?? '',
      sellerName: json['sellerName'] ?? '',
      purchaserName: json['purchaserName'] ?? '',
      commodityDetails: (json['commodityDetails'] as List?)
              ?.map((item) => CommodityDetail.fromJson(item))
              .toList() ??
          [],
    );
  }
}

/// 商品详情
class CommodityDetail {
  final String name;
  final String amount;
  final String price;

  CommodityDetail({
    required this.name,
    required this.amount,
    required this.price,
  });

  factory CommodityDetail.fromJson(Map<String, dynamic> json) {
    return CommodityDetail(
      name: json['name'] ?? '',
      amount: json['amount'] ?? '',
      price: json['price'] ?? '',
    );
  }
}

/// 维修信息
class RepairInfo {
  final List<double> amounts;
  final List<String> dates;
  final List<String> repairItems;
  final double confidence;

  RepairInfo({
    required this.amounts,
    required this.dates,
    required this.repairItems,
    required this.confidence,
  });

  factory RepairInfo.fromJson(Map<String, dynamic> json) {
    return RepairInfo(
      amounts: (json['amounts'] as List?)
              ?.map((item) => (item as num).toDouble())
              .toList() ??
          [],
      dates: List<String>.from(json['dates'] ?? []),
      repairItems: List<String>.from(json['repairItems'] ?? []),
      confidence: (json['confidence'] ?? 0).toDouble(),
    );
  }
}

/// OCR配置
class OCRConfig {
  final int maxFileSize;
  final List<String> allowedMimeTypes;
  final List<String> supportedTypes;

  OCRConfig({
    required this.maxFileSize,
    required this.allowedMimeTypes,
    required this.supportedTypes,
  });

  factory OCRConfig.fromJson(Map<String, dynamic> json) {
    return OCRConfig(
      maxFileSize: json['maxFileSize'] ?? 5242880,
      allowedMimeTypes: List<String>.from(json['allowedMimeTypes'] ?? []),
      supportedTypes: List<String>.from(json['supportedTypes'] ?? []),
    );
  }
}

File: backend/services/ocr-service.js
const AipOcrClient = require('baidu-aip-sdk').ocr;
const fs = require('fs');
const path = require('path');
const sharp = require('sharp');
const config = require('../config');

class OCRService {
  constructor() {
    this.client = new AipOcrClient(
      config.baiduOcr.appId,
      config.baiduOcr.apiKey,
      config.baiduOcr.secretKey
    );
    
    // 设置超时时间
    this.client.timeout = 10000;
  }

  /**
   * 预处理图片 - 提高OCR识别准确率
   */
  async preprocessImage(imagePath) {
    try {
      const processedPath = imagePath.replace(/\.(jpg|jpeg|png)$/i, '_processed.jpg');
      
      await sharp(imagePath)
        .resize(2000, 2000, { 
          fit: 'inside',
          withoutEnlargement: true 
        })
        .sharpen()
        .normalize()
        .jpeg({ quality: 90 })
        .toFile(processedPath);
        
      return processedPath;
    } catch (error) {
      console.error('图片预处理失败:', error);
      return imagePath; // 返回原图片路径
    }
  }

  /**
   * 通用OCR文字识别
   */
  async recognizeText(imagePath, options = {}) {
    try {
      const processedImagePath = await this.preprocessImage(imagePath);
      const image = fs.readFileSync(processedImagePath);
      
      const ocrOptions = {
        language_type: 'CHN_ENG',
        detect_direction: true,
        detect_language: true,
        probability: true,
        ...options
      };

      const result = await this.client.generalBasic(image, ocrOptions);
      
      // 清理临时文件
      if (processedImagePath !== imagePath && fs.existsSync(processedImagePath)) {
        fs.unlinkSync(processedImagePath);
      }
      
      if (result.error_code) {
        throw new Error(`OCR识别失败: ${result.error_msg}`);
      }

      return this.formatGeneralResult(result);
    } catch (error) {
      console.error('OCR识别失败:', error);
      throw error;
    }
  }

  /**
   * 车牌号识别
   */
  async recognizeLicensePlate(imagePath) {
    try {
      const processedImagePath = await this.preprocessImage(imagePath);
      const image = fs.readFileSync(processedImagePath);
      
      const result = await this.client.licensePlate(image);
      
      // 清理临时文件
      if (processedImagePath !== imagePath && fs.existsSync(processedImagePath)) {
        fs.unlinkSync(processedImagePath);
      }
      
      if (result.error_code) {
        throw new Error(`车牌识别失败: ${result.error_msg}`);
      }

      return this.formatLicensePlateResult(result);
    } catch (error) {
      console.error('车牌识别失败:', error);
      throw error;
    }
  }

  /**
   * VIN码识别
   */
  async recognizeVIN(imagePath) {
    try {
      const result = await this.recognizeText(imagePath, {
        language_type: 'ENG',
        detect_direction: true
      });
      
      // 从识别结果中提取可能的VIN码
      const vinPattern = /[A-HJ-NPR-Z0-9]{17}/g;
      const extractedText = result.words.map(item => item.words).join(' ');
      const vinMatches = extractedText.match(vinPattern);
      
      return {
        success: true,
        vins: vinMatches || [],
        fullText: extractedText,
        confidence: result.words.length > 0 ? 
          result.words.reduce((sum, item) => sum + item.probability, 0) / result.words.length : 0
      };
    } catch (error) {
      console.error('VIN码识别失败:', error);
      throw error;
    }
  }

  /**
   * 发票信息识别
   */
  async recognizeInvoice(imagePath) {
    try {
      const processedImagePath = await this.preprocessImage(imagePath);
      const image = fs.readFileSync(processedImagePath);
      
      // 先尝试增值税发票识别
      let result;
      try {
        result = await this.client.vatInvoice(image);
      } catch (vatError) {
        // 如果增值税发票识别失败，使用通用发票识别
        result = await this.client.receipt(image);
      }
      
      // 清理临时文件
      if (processedImagePath !== imagePath && fs.existsSync(processedImagePath)) {
        fs.unlinkSync(processedImagePath);
      }
      
      if (result.error_code) {
        throw new Error(`发票识别失败: ${result.error_msg}`);
      }

      return this.formatInvoiceResult(result);
    } catch (error) {
      console.error('发票识别失败:', error);
      throw error;
    }
  }

  /**
   * 格式化通用OCR结果
   */
  formatGeneralResult(result) {
    return {
      success: true,
      direction: result.direction || 0,
      language: result.language || 'unknown',
      words: (result.words_result || []).map(item => ({
        words: item.words,
        probability: item.probability || 0,
        location: item.location || null
      })),
      fullText: (result.words_result || []).map(item => item.words).join('\n')
    };
  }

  /**
   * 格式化车牌识别结果
   */
  formatLicensePlateResult(result) {
    const wordsResult = result.words_result;
    return {
      success: true,
      plateNumber: wordsResult?.number || '',
      color: wordsResult?.color || '',
      confidence: wordsResult?.probability || 0,
      location: wordsResult?.vertexes_location || null
    };
  }

  /**
   * 格式化发票识别结果
   */
  formatInvoiceResult(result) {
    const wordsResult = result.words_result || {};
    
    return {
      success: true,
      invoiceType: wordsResult.InvoiceType || '',
      invoiceCode: wordsResult.InvoiceCode || '',
      invoiceNum: wordsResult.InvoiceNum || '',
      invoiceDate: wordsResult.InvoiceDate || '',
      totalAmount: wordsResult.TotalAmount || '',
      amountInWords: wordsResult.AmountInWords || '',
      sellerName: wordsResult.SellerName || '',
      purchaserName: wordsResult.PurchaserName || '',
      commodityDetails: this.extractCommodityDetails(wordsResult),
      fullResult: wordsResult
    };
  }

  /**
   * 提取商品详情
   */
  extractCommodityDetails(wordsResult) {
    const details = [];
    
    // 处理不同类型发票的商品信息
    if (wordsResult.CommodityName) {
      details.push({
        name: wordsResult.CommodityName,
        amount: wordsResult.CommodityAmount || '',
        price: wordsResult.CommodityPrice || ''
      });
    }
    
    return details;
  }

  /**
   * 智能提取维修相关信息
   */
  extractRepairInfo(ocrResult) {
    const fullText = ocrResult.fullText || '';
    const words = ocrResult.words || [];
    
    // 提取金额
    const amountPattern = /(?:金额|费用|总计|合计|应付)[：:\s]*¥?(\d+(?:\.\d{2})?)/gi;
    const amounts = [];
    let match;
    while ((match = amountPattern.exec(fullText)) !== null) {
      amounts.push(parseFloat(match[1]));
    }
    
    // 提取日期
    const datePattern = /(\d{4}[-年]\d{1,2}[-月]\d{1,2}[日]?)/g;
    const dates = [];
    while ((match = datePattern.exec(fullText)) !== null) {
      dates.push(match[1].replace(/[年月]/g, '-').replace(/日/g, ''));
    }
    
    // 提取维修项目关键词
    const repairKeywords = [
      '换油', '保养', '维修', '更换', '检查', '清洗', '调整', '修理',
      '机油', '刹车', '轮胎', '空调', '电池', '火花塞', '滤芯'
    ];
    
    const detectedItems = [];
    repairKeywords.forEach(keyword => {
      if (fullText.includes(keyword)) {
        detectedItems.push(keyword);
      }
    });
    
    return {
      amounts: amounts,
      dates: dates,
      repairItems: detectedItems,
      confidence: ocrResult.words.length > 0 ? 
        ocrResult.words.reduce((sum, item) => sum + item.probability, 0) / ocrResult.words.length : 0
    };
  }
}

module.exports = new OCRService(); 